harvester/DBInterface.py

```diff
@@ -77,107 +69,199 @@ def _prep(self, statement):
 		if record == None:
 			return None
+		record["record_id"] = self.get_record_id(repo_id, record["identifier"])
 
 		con = self.getConnection()
 		with con:
 			cur = con.cursor()
-			verb = "INSERT"
-			if mode == "replace":
-				verb = "REPLACE"
+			source_url = ""
+			if 'dc:source' in record:
+				source_url = record["dc:source"]
+			if record["record_id"] == 0:
+				try:
+					if self.dbtype == "postgres":
+						cur.execute(self._prep("INSERT INTO records (title, pub_date, contact, series, modified_timestamp, source_url, deleted, local_identifier, repository_id) VALUES(?,?,?,?,?,?,?,?,?) RETURNING record_id"), 
+							(record["title"], record["pub_date"], record["contact"], record["series"], time.time(), source_url, 0, record["identifier"], repo_id))
+						res = cur.fetchone()
+						record["record_id"] = int(res[0])
+					if self.dbtype == "sqlite":
+						cur.execute(self._prep("INSERT INTO records (title, pub_date, contact, series, modified_timestamp, source_url, deleted, local_identifier, repository_id) VALUES(?,?,?,?,?,?,?,?,?)"), 
+							(record["title"], record["pub_date"], record["contact"], record["series"], time.time(), source_url, 0, record["identifier"], repo_id))
+						record["record_id"] = int(cur.lastrowid)
+				except self.dblayer.IntegrityError:
+					return None
+			else:
+				cur.execute(self._prep("UPDATE records set title=?, pub_date=?, contact=?, series=?, modified_timestamp=?, source_url=?, deleted=?, local_identifier=? WHERE record_id = ?"),
+					(record["title"], record["pub_date"], record["contact"], record["series"], time.time(), source_url, 0, record["identifier"], record["record_id"]) )
 
-			try:
-				if 'dc:source' in record:
-					cur.execute(self._prep(verb + " INTO records (title, date, contact, series, modified_timestamp, source_url, deleted, local_identifier, repository_url) VALUES(?,?,?,?,?,?,?,?,?)"), (record["title"], record["date"], record["contact"], record["series"], time.time(), record["dc:source"], 0, record["identifier"], repository_url))
-				else:
-					cur.execute(self._prep(verb + " INTO records (title, date, contact, series, modified_timestamp, deleted, local_identifier, repository_url) VALUES(?,?,?,?,?,?,?,?)"), (record["title"], record["date"], record["contact"], record["series"], time.time(), 0, record["identifier"], repository_url))
-			except self.dblayer.IntegrityError:
-				# record already present in repo
+			if record["record_id"] == 0:
 				return None
 
 			if "creator" in record:
 				if not isinstance(record["creator"], list):
 					record["creator"] = [record["creator"]]
 				for creator in record["creator"]:
 					try:
-						cur.execute(self._prep("INSERT INTO creators (local_identifier, repository_url, creator, is_contributor) VALUES (?,?,?,?)"), (record["identifier"], repository_url, creator, 0))
+						cur.execute(self._prep("DELETE from creators where record_id = ? and is_contributor=0"), (record["record_id"],))
```

Comment:
Is this approach necessary? Deleting and removing the corresponding rows on every run?